### PR TITLE
Test/gam/playwright e2e suite for HU-GAM-02

### LINKED ISSUE #179: TEST-GAM-02 - Earn points end-to-end test report
# TEST-GAM-02 – Earn points end-to-end test report

**Domain:** Gamification

## User Story

As an authenticated ITESO user, I want to earn points when I complete marketplace actions so that I feel rewarded and motivated to keep using the platform.

## Descripción detallada

This test report covers end-to-end and edge-case scenarios for the earn-points feature implemented in issue #16 (HU-GAM-02). The feature awards points to users when they perform marketplace actions using the `award-points` endpoint. Points are stored in the `User.points` field and reflected in the `PointsBalance` component on the `/profile` page.

Reference issue: https://github.com/ReUseITESO/ReUse/issues/16

## Acceptance Criteria

- Balance de puntos visible en el perfil del usuario autenticado
- El balance refleja el valor real reportado por `/api/gamification/level-progression/`
- `award-points` retorna 200 para todas las acciones válidas del marketplace (`publish_item`, `complete_donation`, `complete_sale`, `complete_exchange`, `receive_positive_review`)
- El balance del usuario aumenta tras llamar `award-points`
- La sección de actividad reciente muestra las últimas transacciones con signo `+`
- Estado de carga (skeleton) visible mientras se obtienen los datos
- Error de API muestra botón de Reintentar
- Usuario no autenticado ve aviso apropiado; rutas protegidas retornan 401
- La visualización es correcta en viewport móvil (375px) y escritorio (1280px)

### Backend

- `POST /api/gamification/award-points/` — awards points for a given action and user
- `GET /api/gamification/points/` — returns `{ points: number }` for authenticated user
- `GET /api/gamification/level-progression/` — returns points, current_level, next_level, progress_percent
- `GET /api/gamification/points/history/` — paginated transactions with action, points, created_at

### Frontend

- `PointsBalance` component in `src/components/gamification/PointsBalance.tsx` — shows total points, level progression gradient, and up to 3 recent transactions
- `/profile` page — main page where balance and recent activity are displayed
- `/profile/points-history` page — full paginated history via `PointsHistoryCard`

### Database

- `User.points` integer field updated on every award
- `PointTransaction` records created per award call

## Testing Notes

Test strategy:
- Automated E2E tests with Playwright covering API contract, UI display, auth protection, loading/error states, and responsiveness.
- Tests use a real backend (Docker) and seed data; no full mocking except for error-state simulations.
- Test user: `jose.chavez@iteso.mx / ReUse2026!`
- Prerequisite: `python manage.py seed_dev_data`

## Test Scenarios (Playwright)

1) **AC1 – Balance visible en el perfil**
   - Usuario autenticado navega a `/profile`
   - Verificar que la sección "Puntos Acumulados" existe y muestra un número
   - Verificar que el número coincide con el valor de `/api/gamification/level-progression/`
   - Verificar que el link "Ver todo el historial" apunta a `/profile/points-history`

2) **AC2 – award-points acepta todas las acciones válidas**
   - Llamar `POST /api/gamification/award-points/` con cada una de las 5 acciones
   - Verificar que cada llamada retorna HTTP 200 con `{ message: "Points awarded successfully" }`
   - Verificar que llamada sin auth retorna 401
   - Verificar que acción inválida retorna 400

3) **AC3 – Balance aumenta después de ganar puntos**
   - Obtener puntos antes via `GET /api/gamification/points/`
   - Llamar `award-points` con `publish_item`
   - Obtener puntos después y verificar incremento
   - Verificar estructura completa de `/api/gamification/level-progression/`

4) **AC4 – Actividad reciente**
   - Verificar que la sección "Actividad reciente" existe en `/profile`
   - Después de ganar puntos, verificar que aparece una entrada con texto de acción y signo `+`
   - Verificar estructura de `/api/gamification/points/history/` (results array, campos id, action, points, created_at)

5) **AC5 – Estado de carga y error**
   - Simular delay en `level-progression`: verificar que skeleton `.animate-pulse` aparece
   - Simular 500 en `level-progression`: verificar que aparece botón "Reintentar"

6) **AC6 – Protección para usuario no autenticado**
   - Navegar a `/profile` sin token: verificar aviso de login
   - Llamar `GET /api/gamification/points/` sin token: verificar 401
   - Llamar `GET /api/gamification/level-progression/` sin token: verificar 401
   - Llamar `GET /api/gamification/points/history/` sin token: verificar 401

7) **AC7 – Responsividad**
   - Cargar `/profile` en 1280x800: verificar tarjeta de puntos visible
   - Cargar `/profile` en 375x812: verificar tarjeta visible y dentro del viewport
   - Verificar que el link al historial funciona en móvil

## Test Data Required

- Usuario seed: `jose.chavez@iteso.mx / ReUse2026!`
- PointRules activas para las 5 acciones del marketplace (creadas por seed)
- Al menos 1 PointTransaction para verificar actividad reciente

## Potential Mocks

- `GET /api/gamification/level-progression/` con delay para probar skeleton
- `GET /api/gamification/level-progression/` con 500 para probar estado de error
- No se mockea la DB; los tests de API usan el backend real

## Dependencies & Duplication Check

**Dependencies:** HU-GAM-02 (implementación de earn-points, issue #16)

**Possible duplicates:** HU-GAM-02A (point actions, issue #77) — cubierto en TEST-GAM-02A

## Assumptions

- Backend corriendo en `http://localhost:8000`
- Frontend corriendo en `http://localhost:3000`
- Seed ejecutado: `python manage.py seed_dev_data`
- Token se inyecta via `localStorage` (`reuse_access_token`, `reuse_refresh_token`)

## ⚠️ Warnings

- Si las PointRules no están seedeadas, los tests de AC2/AC3 fallarán porque `award-points` no encontrará la regla activa.
- El balance mostrado puede diferir si otro test ya modificó los puntos del usuario seed; los tests de balance son relativos (mayor que antes), no absolutos.

### PR DESCRIPTION
**Pull request by OmarTieso**
## Summary

- Adds Playwright E2E test suite for HU-GAM-02 (earn points)
- Covers AC1–AC7: points balance on profile, award-points API for all marketplace actions, balance update, recent activity, loading/error states, auth protection, and responsive layout
- Adds `playwright.config.ts` and `test:e2e` scripts to `package.json`

Closes #179

## Test plan

- [ ] Backend running at `http://localhost:8000`
- [ ] Frontend running at `http://localhost:3000`
- [ ] Seed executed: `python manage.py seed_dev_data`
- [ ] Run: `cd frontend && npx playwright test e2e/earn-points-profile.spec.ts`
- [ ] Verify all ACs pass; document any failures in issue #179

### CODE REVIEW

**Comment by Ferreirafc1133** [comment]
omar, dev ya tiene playwright.config en tests/gamification/ desde que #164 mergeo. tu estas armando uno nuevo en e2e/ con webServer como si nadie hubiera ido primero. rebasea, borra tu config, borra los scripts del package.json (ya estan) y mueve el spec a tests/gamification/.

File: frontend/playwright.config.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { defineConfig, devices } from '@playwright/test';
+
+export default defineConfig({
+  testDir: './e2e',
```
Comment (Ferreirafc1133) [inline frontend/playwright.config.ts]: testDir e2e/ no, dev ya tiene tests/gamification/ desde #164. rebasea y borra este config.

**Comment by Ferreirafc1133** [comment]
@OmarTieso ping. dos dias sin rebase. el config de tests/gamification/ ya esta en dev.

**Comment by Ferreirafc1133** [comment]
@OmarTieso tercer ping. dev tiene playwright config en tests/gamification/ desde hace 5 dias. se nota que no rebaseas porque tu PR sigue trayendo cosas que ya viven en dev.

**Comment by Ferreirafc1133** [comment]
@OmarTieso esta PR esta en conflicto. rebasea contra dev y unifica con la estructura tests/gamification/ que ya quedo.

**Comment by Ferreirafc1133** [comment]
@OmarTieso 6 dias sin movimiento. dev ya tiene tests/gamification/ con su config. rebasea o la cierro.

**Comment by OmarTieso** [comment]
Mismas correcciones que #181. Además eliminé el earn-points-profile.spec.ts que estaba duplicado con #181. Aproveché para corregir un bug detectado por los tests: badge_service.py otorgaba puntos por badges sin crear PointTransaction, causando inconsistencia entre el total de puntos y el historial.
